### Овсянникова Майя группа 241, 03.02.2026

### Условие задачи
1. Создать новые атрибуты: `year`, `month`, `quarter`, `day_of_week`, `day_of_year`, `is_month_end`, `season`: (`'Winter'`, `'Spring'`, `'Summer'`, `'Autumn'`)

2. Определить самый прибыльный день недели по средней прибыли и самый загруженный день по количеству заказов

3. Выделить данные только за выходные дни (суббота и воскресенье) и сравните средний чек с рабочими днями

4. Сгруппировать данные по кварталам и подсчитать для каждого квартала:
 - a. общий объем продаж
 - b. среднюю прибыльность
 - c. количество уникальных клиентов
 - d. определите лучший квартал по каждому из показателей

### Теория
**0. Подготовка данных:**
- Загрузила csv-файл с kaggle
- Проверила наличие пропусков и удалила строки с `NaN`
- Преобразовала столбец с датами `df['Order Date']` в формат `datetime` с помощью `pd.to_datetime()`

**1. Создание yовых временных атрибутов:**
- Для создания новых атрибутов использовала методы `.dt` pandas:
    - `df['Order Date'].dt.year` – год
    - `df['Order Date'].dt.month` – месяц (число)
    - `df['Order Date'].dt.month_name()` – месяц (название)
    - `df['Order Date'].dt.quarter` – квартал
    - `df['Order Date'].dt.dayofweek` – день недели (число 0-6)
    - `df['Order Date'].dt.day_name()` – день недели (название)
    - `df['Order Date'].dt.dayofyear` – день года
    - `df['Order Date'].dt.is_month_end` – является ли последним днем месяца
- Для определения сезона применила лямбда-функцию через
- Добавила вспомогательные числовые колонки month_num и day_of_week_num для удобства расчетов

**2. Определение самого прибыльного и загруженного дня недели:**
- Группировала данные по дням недели
- Для прибыльности считала среднее значение прибыли: `df.groupby('day_of_week')['Profit'].mean()`
- Для загруженности считала количество уникальных заказов: `df.groupby('day_of_week')['Order ID'].nunique()`
- Нашла сам день и максимальное искомое значение с помощью `.idxmax()` и `.max()`

**3. Сравнение выручки выходных и рабочих дней:**
- Создала фильтры: `df['day_of_week_num'] >= 5` (выходные) и `df['day_of_week_num'] < 5` (рабочие дни)
- Для каждой группы посчитала средний чек (сумму продаж по заказу) `df[фильтр].groupby('Order ID')['Sales'].sum().mean()`
- Сравнила полученные результаты

**4. Анализ кварталов:**
- Группировала данные по году и кварталу `df.groupby(['year', 'quarter'])`
- Рассчитала показатели:
    - `df.groupby(['year', 'quarter'])['Sales'].sum()` – общий объем продаж
    - `df.groupby(['year', 'quarter'])['Profit'].mean()` – средняя прибыльность
    - `df.groupby(['year', 'quarter'])['Customer ID'].nunique()` – количество уникальных клиентов
- Для каждого показателя определила лучший квартал с помощью `.idxmax()`

### Документация

Используемые библиотеки:
- pandas – основная библиотека для работы с табличными данными

Основные методы pandas:
- `pd.read_csv()` – загрузка данных из csv файла
- `df.dropna()` – удаление строк с пропущенными значениями
- `pd.to_datetime()` – преобразование строк в формат даты
- `df.sort_values()` – сортировка данных по указанному столбцу
- `.dt` – доступ к компонентам даты:
    - `.dt.year`, `.dt.month`, `.month_name()`, `.dt.quarter` – год, месяц, название месяца квартал
    - `.dt.dayofweek`, `.dt.day_name()` – день недели (число и название)
    - `.dt.dayofyear` – день года
    - `.dt.is_month_end` – проверка последнего дня месяца
- `df.groupby()` – группировка данных для агрегации
    -`.mean()`, `.sum()`, `.nunique()` – функции агрегации (среднее, сумма, количество уникальных)
- `.idxmax()` – поиск индекса максимального значения
- `.apply()` с лямбда-функцией – применение функции к каждому элементу

### Решение

#### Загрузка и предварительная обработка данных

In [ ]:
import pandas as pd

df = pd.read_csv('samplesuperstore.csv', sep=',', encoding='cp1252')
df = df.dropna()

print(df.shape)

(9994, 21)


#### 1. Создать новые атрибуты:
`year`, `month`, `quarter`, `day_of_week`, `day_of_year`, `is_month_end`, `season`: (`'Winter'`, `'Spring'`, `'Summer'`, `'Autumn'`)

In [5]:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df = df.sort_values('Order Date')


df['year'] = df['Order Date'].dt.year
df['month_num'] = df['Order Date'].dt.month
df['month'] = df['Order Date'].dt.month_name()
df['quarter'] = df['Order Date'].dt.quarter
df['day_of_week_num'] = df['Order Date'].dt.dayofweek
df['day_of_week'] = df['Order Date'].dt.day_name()
df['day_of_year'] = df['Order Date'].dt.dayofyear
df['is_month_end'] = df['Order Date'].dt.is_month_end
df['season'] = df['month_num'].apply(lambda m: 'Winter' if m in [12, 1, 2] else
                                               'Spring' if m in [3, 4, 5] else
                                               'Summer' if m in [6, 7, 8] else
                                               'Autumn')

print("Созданные атрибуты (без воспомогательных):")
cols = ['Order Date', 'year', 'month', 'quarter', 'day_of_week', 'day_of_year', 'is_month_end', 'season']
display(df[cols].head(10))

print("\nis_month_end = True:")
display(df[df['is_month_end']][cols].head(5))

Созданные атрибуты (без воспомогательных):


,Order Date,year,month,quarter,day_of_week,day_of_year,is_month_end,season
7980,2014-01-03,2014,January,1,Friday,3,False,Winter
739,2014-01-04,2014,January,1,Saturday,4,False,Winter
740,2014-01-04,2014,January,1,Saturday,4,False,Winter
741,2014-01-04,2014,January,1,Saturday,4,False,Winter
1759,2014-01-05,2014,January,1,Sunday,5,False,Winter
7476,2014-01-06,2014,January,1,Monday,6,False,Winter
7474,2014-01-06,2014,January,1,Monday,6,False,Winter
7475,2014-01-06,2014,January,1,Monday,6,False,Winter
7180,2014-01-06,2014,January,1,Monday,6,False,Winter
7477,2014-01-06,2014,January,1,Monday,6,False,Winter



is_month_end = True:


,Order Date,year,month,quarter,day_of_week,day_of_year,is_month_end,season
3365,2014-01-31,2014,January,1,Friday,31,True,Winter
9566,2014-03-31,2014,March,1,Monday,90,True,Spring
5479,2014-03-31,2014,March,1,Monday,90,True,Spring
9567,2014-03-31,2014,March,1,Monday,90,True,Spring
4711,2014-03-31,2014,March,1,Monday,90,True,Spring


#### 2. Определить самый прибыльный день недели по средней прибыли и самый загруженный день по количеству заказов

In [7]:
most_profit_day = df.groupby('day_of_week')['Profit'].mean()
most_order_day = df.groupby('day_of_week')['Order ID'].nunique()


print("most_profit_day:", most_profit_day.idxmax(), "– avg profit:", most_profit_day.max().round(2))
print("most_order_day:", most_order_day.idxmax(), "– orders:", most_order_day.max())

most_profit_day: Wednesday – avg profit: 39.63
most_order_day: Monday – orders: 920


#### 3. Выделить данные только за выходные дни (суббота и воскресенье) и сравните средний чек с рабочими днями

In [76]:
avg_weekend = df[df['day_of_week_num'] >= 5].groupby('Order ID')['Sales'].sum().mean().round(2)
avg_workday = df[df['day_of_week_num'] < 5].groupby('Order ID')['Sales'].sum().mean().round(2)

print("avg_weekend:", avg_weekend)
print("avg_workday:", avg_workday)

avg_weekend: 440.63
avg_workday: 467.75


#### 4. Сгруппировать данные по кварталам и подсчитать для каждого квартала:
 - a. общий объем продаж
 - b. среднюю прибыльность
 - c. количество уникальных клиентов
 - d. определите лучший квартал по каждому из показателей

In [88]:
task4a = df.groupby(['year', 'quarter'])['Sales'].sum().round(2)
task4b = df.groupby(['year', 'quarter'])['Profit'].mean().round(2)
task4c = df.groupby(['year', 'quarter'])['Customer ID'].nunique()


print("\na: общий объем продаж\n")
display(task4a)

print("\nb: средняя прибыльность")
display(task4b)

print("\nc: количество уникальных клиентов")
display(task4c)

print("\nd_a: лучший квартал по общему объему продаж:", task4a.idxmax()[0], "Q", task4a.idxmax()[1],
      "\nd_b: лучший квартал по средней прибыли:",  task4b.idxmax()[0], "Q", task4b.idxmax()[1],
      "\nd_c: лучший квартал по количеству уникальных клиентов:", task4c.idxmax()[0], "Q", task4c.idxmax()[1])


a: общий объем продаж



year  quarter
2014  1           74447.80
      2           86538.76
      3          143633.21
      4          179627.73
2015  1           68851.74
      2           89124.19
      3          130259.58
      4          182297.01
2016  1           93237.18
      2          136082.30
      3          143787.36
      4          236098.75
2017  1          123144.86
      2          133764.37
      3          196251.96
      4          280054.07
Name: Sales, dtype: float64


b: средняя прибыльность


year  quarter
2014  1          13.51
      2          28.58
      3          22.70
      4          28.77
2015  1          35.63
      2          27.46
      3          28.47
      4          28.92
2016  1          34.15
      2          27.59
      3          21.38
      4          41.55
2017  1          47.01
      2          22.46
      3          29.88
      4          22.52
Name: Profit, dtype: float64


c: количество уникальных клиентов


year  quarter
2014  1          121
      2          180
      3          228
      4          305
2015  1          135
      2          192
      3          233
      4          310
2016  1          162
      2          240
      3          300
      4          345
2017  1          218
      2          288
      3          343
      4          436
Name: Customer ID, dtype: int64


d_a: лучший квартал по общему объему продаж: 2017 Q 4 
d_b: лучший квартал по средней прибыли: 2017 Q 1 
d_c: лучший квартал по количеству уникальных клиентов: 2017 Q 4


### Вывод
Получены ответы на все пункты

**Задание 1:** Созданы 9 новых атрибутов, которые используются в следующих заданиях:
 - 7 требуемых (`year`, `month`, `quarter`, `day_of_week`, `day_of_year`, `is_month_end`, `season`)
 - 2 воспомогательных (`month_num`, `day_of_week_num`)


**Задание 2:** Самый прибыльный день недели – `Wednesday` (среда) – средняя прибыль `39.63`

Самый загруженный день – `Monday` (понедельник) – `920` уникальных заказов



**Задание 3:** Средний чек в рабочие дни `467.75` выше, чем в выходные `440.63`

Разница составляет `27.12`, что говорит о более крупных покупках в рабочие дни


**Задание 4:** Получена поквартальная статистика по объемам продаж, средней прибыльности и колоичеству уникальных клиентов
 - лучший квартал по общему объему продаж: `2017 Q4 `
 - лучший квартал по средней прибыли: `2017 Q1 `
 - лучший квартал по количеству уникальных клиентов: `2017 Q4`
